### PR TITLE
Update typing for Python 3.7 (1)

### PR DESCRIPTION
**Pull request by cdce8p**
## Description
Part 1 of 3 to update the typing in astroid following the bump to Python 3.7.2

Part 1: Replace `typing` aliases for builtin types
Part 2: Replace `typing` aliases for `collections.abc` classes.
Part 3: Update `py-version` in `pylintrc`

### CODE REVIEW

File: astroid/brain/brain_builtin_inference.py [hunk 1]
```diff
@@ -3,10 +3,11 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Astroid hooks for various builtins."""
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/brain/brain_builtin_inference.py:6]:
```suggestion

from __future__ import annotations
```

File: astroid/brain/brain_namedtuple_enum.py [hunk 2]
```diff
@@ -3,11 +3,12 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Astroid hooks for the Python standard library."""
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/brain/brain_namedtuple_enum.py:6]:
```suggestion

from __future__ import annotations
```

File: astroid/filter_statements.py [hunk 3]
```diff
@@ -5,15 +5,14 @@
 """_filter_stmts and helper functions. This method gets used in LocalsDictnodes.NodeNG._scope_lookup.
 It is not considered public.
 """
-
-from typing import List, Optional, Tuple
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/filter_statements.py:8]:
```suggestion

from __future__ import annotations
```

File: astroid/inference.py [hunk 4]
```diff
@@ -4,24 +4,13 @@
 
 """this module contains a set of functions to handle inference on astroid trees
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/inference.py:7]:
```suggestion

from __future__ import annotations
```

File: astroid/inference_tip.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Transform utilities (filters and decorator)"""
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/inference_tip.py:6]:
```suggestion

from __future__ import annotations
```

File: astroid/interpreter/objectmodel.py [hunk 6]
```diff
@@ -20,13 +20,14 @@
 attribute. Thus the model can be viewed as a special part of the lookup
 mechanism.
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/interpreter/objectmodel.py:23]:
```suggestion

from __future__ import annotations
```

File: astroid/manager.py [hunk 7]
```diff
@@ -6,12 +6,13 @@
 possible by providing a class responsible to get astroid representation
 from various source and using a cache of built modules)
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/manager.py:9]:
```suggestion

from __future__ import annotations
```

File: astroid/mixins.py [hunk 8]
```diff
@@ -4,9 +4,11 @@
 
 """This module contains some mixins for the different nodes.
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/mixins.py:7]:
```suggestion

from __future__ import annotations
```

File: astroid/modutils.py [hunk 9]
```diff
@@ -13,6 +13,7 @@
 :type BUILTIN_MODULES: dict
 :var BUILTIN_MODULES: dictionary with builtin module names has key
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/modutils.py:16]:
```suggestion

from __future__ import annotations
```

File: astroid/nodes/node_classes.py [hunk 10]
```diff
@@ -3,6 +3,7 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Module for some node classes. More nodes in scoped_nodes.py"""
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/nodes/node_classes.py:6]:
```suggestion

from __future__ import annotations
```

File: astroid/nodes/scoped_nodes/mixin.py [hunk 11]
```diff
@@ -3,8 +3,9 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """This module contains mixin classes for scoped nodes."""
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/nodes/scoped_nodes/mixin.py:6]:
```suggestion

from __future__ import annotations
```

File: astroid/nodes/scoped_nodes/scoped_nodes.py [hunk 12]
```diff
@@ -7,23 +7,15 @@
 new local scope in the language definition : Module, ClassDef, FunctionDef (and
 Lambda, GeneratorExp, DictComp and SetComp to some extent).
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/nodes/scoped_nodes/scoped_nodes.py:10]:
```suggestion

from __future__ import annotations
```

File: astroid/nodes/scoped_nodes/utils.py [hunk 13]
```diff
@@ -5,20 +5,21 @@
 """
 This module contains utility functions for scoped nodes.
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/nodes/scoped_nodes/utils.py:8]:
```suggestion

from __future__ import annotations
```

File: astroid/objects.py [hunk 14]
```diff
@@ -10,9 +10,10 @@
 
     Call(func=Name('frozenset'), args=Tuple(...))
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/objects.py:13]:
```suggestion

from __future__ import annotations
```

File: astroid/protocols.py [hunk 15]
```diff
@@ -5,11 +5,12 @@
 """this module contains a set of functions to handle python protocols for nodes
 where it makes sense.
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/protocols.py:8]:
```suggestion

from __future__ import annotations
```

File: astroid/rebuilder.py [hunk 16]
```diff
@@ -5,26 +5,14 @@
 """this module contains utilities for rebuilding an _ast tree in
 order to get a single Astroid representation
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/rebuilder.py:8]:
```suggestion

from __future__ import annotations
```

File: astroid/test_utils.py [hunk 17]
```diff
@@ -3,11 +3,13 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Utility functions for test code that uses astroid ASTs as input."""
+from __future__ import annotations
```
Comment (DanielNoord) [inline astroid/test_utils.py:6]:
```suggestion

from __future__ import annotations
```

File: pylintrc [hunk 18]
```diff
@@ -310,7 +310,7 @@ mixin-class-rgx=.*Mix[Ii]n
 # List of members which are set dynamically and missed by pylint inference
 # system, and so shouldn't trigger E0201 when accessed. Python regular
 # expressions are accepted.
-generated-members=REQUEST,acl_users,aq_parent,argparse.Namespace
+generated-members=REQUEST,acl_users,aq_parent,argparse.Namespace,ast\.([mM]atch.*|pattern)
```
Comment (DanielNoord) [inline pylintrc:313]: ? Shouldn't this be in a different PR/commit?

File: script/bump_changelog.py [hunk 19]
```diff
@@ -6,12 +6,13 @@
 This script permits to upgrade the changelog in astroid or pylint when releasing a version.
 """
 # pylint: disable=logging-fstring-interpolation
+from __future__ import annotations
```
Comment (DanielNoord) [inline script/bump_changelog.py:9]:
```suggestion

from __future__ import annotations
```

File: tests/unittest_brain.py [hunk 20]
```diff
@@ -3,12 +3,14 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Tests for basic functionality in astroid.brain."""
+from __future__ import annotations
```
Comment (DanielNoord) [inline tests/unittest_brain.py:6]:
```suggestion

from __future__ import annotations
```

File: tests/unittest_inference.py [hunk 21]
```diff
@@ -3,13 +3,14 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Tests for the astroid inference capabilities"""
+from __future__ import annotations
```
Comment (DanielNoord) [inline tests/unittest_inference.py:6]:
```suggestion

from __future__ import annotations
```

File: tests/unittest_nodes.py [hunk 22]
```diff
@@ -4,12 +4,14 @@
 
 """tests for specific behaviour of astroid nodes
 """
+from __future__ import annotations
```
Comment (DanielNoord) [inline tests/unittest_nodes.py:7]:
```suggestion

from __future__ import annotations
```

**Review by DanielNoord — COMMENTED** [review]
I added a white line between imports and docstring whenever there originally was one.

Another approach would be to `replace all` on the `__future__` imports with a white lines in front of it. `Black` should then remove those case where there are more than one line and we would have consitency across all files.

Rest LGTM!

File: pylintrc (hunk 18 above)
Comment (cdce8p) [reply pylintrc:313]: Need this to fix a `no-member` issue now that the string annotations are replaced with `from __future__ import annotations`, see https://github.com/PyCQA/pylint/issues/6594.

Comment (DanielNoord) [reply pylintrc:313]: Ah! That explains it!

**Comment by cdce8p** [comment]
> I added a white line between imports and docstring whenever there originally was one.

Added empty lines between every docstring and future import (in the changed files). There is an `isort` setting [`lines-before-imports`](https://pycqa.github.io/isort/docs/configuration/options.html#lines-before-imports). Unfortunately though it doesn't seem to work reliably for our use case. More promising is a black PR to make that the default behavior: https://github.com/psf/black/pull/2996.
